depmapomics/copynumbers.py

```diff
@@ -616,8 +618,207 @@ def read_ms_repeats(
+    hgnc_table = hgnc_table.loc[
+        ~(
+            hgnc_table["ensembl_gene_id"].duplicated()
+            | hgnc_table["hugo_entrez"].duplicated()
+        )
+    ]
+
+    # collect genes with multiple locations
+    par_genes = hgnc_table.loc[
+        hgnc_table["location"].str.contains(" and ").fillna(False)
+    ].copy()
+
+    par_genes[["location1", "location2"]] = par_genes["location"].str.split(
+        " and ", expand=True
+    )
+
+    # confirm that we're only looking at PARs here
+    assert bool(par_genes["location1"].str.slice(0, 1).isin(["X", "Y"]).all()) & bool(
+        par_genes["location2"].str.slice(0, 1).isin(["X", "Y"]).all()
+    )
+
+    # indicate all PARs in the table
+    hgnc_table["par"] = hgnc_table["ensembl_gene_id"].isin(par_genes["ensembl_gene_id"])
+    hgnc_table = hgnc_table.drop(columns="location")
+
+    return hgnc_table
+
+
+def aggregate_cnvs_from_hmm(
+    refworkspace,
+    set_entity="sample_set",
+    sampleset="all",
+    gene_level_colname="cnv_cn_by_gene_weighted_mean",
+    seg_level_colname="cnv_segments",
+):
+    """Aggregate gene- and segment-level CN from the HMM-based relative CN pipeline.
+
+    Returns:
+        segments (pd.DataFrame): concatenated long table containing segments.
+            columns: DepMap_ID,Chromosome,Start,End,NumProbes,SegmentMean,Status
+            unlike GATK, this pipeline does not output a "Status", so the column is
+            populated with "."
+        gene_cn_wide (pd.DataFrame): CDS-ID x gene matrix. Columns are "Hugo (Entrez)"
+            gene IDs
+    """
+
+    print("aggregating WGS CN (HMM) from Terra")
+
+    wm = dm.WorkspaceManager(refworkspace)
+
+    sample_ids = pd.Series(
+        [
+            x["entityName"]
+            for x in wm.get_entities(set_entity).loc[sampleset, "samples"]
+        ],
+        name="sample_id",
+    )
+
+    samples = wm.get_samples()
+
+    cnv_files = samples.loc[sample_ids, [gene_level_colname, seg_level_colname]]
+
+    print("Downloading segment files")
+    seg_files = cnv_files[seg_level_colname].dropna()
+    seg_dfs = []
+
+    for sample_id, f in tqdm(seg_files.items(), total=len(seg_files)):
+        seg_dfs.append(
+            pd.read_csv(
+                f,
+                sep="\t",
+                usecols=[
+                    "CONTIG",
+                    "START",
+                    "END",
+                    "LOG2_COPY_RATIO_POSTERIOR_50",
+                    "NUM_POINTS_COPY_RATIO",
+                ],
+                dtype={
+                    "CONTIG": "string",
+                    "START": "int64",
+                    "END": "int64",
+                    "LOG2_COPY_RATIO_POSTERIOR_50": "float64",
+                    "NUM_POINTS_COPY_RATIO": "int64",
+                },
+            ).assign(Sample=sample_id)
+        )
+
+    print("Aggregating segments")
+    segments = pd.concat(seg_dfs, ignore_index=True)
+    del seg_dfs
+    # TODO: drop Y?
```

Comment:
I don't see any reason why we have to drop chrY segments, but there's some logic in the GATK version that does.

Reply:
I think this is a pretty outdated comment. In the old CN pipeline we dropped the Y segments for samples that have more than 150 segments on chrY (indicating noise/poor segmentation), but with the new WGS CN pipeline we're not doing that anymore.